Smol String

milseman opened the pull request:
<!-- What's in this pull request? -->

Add in small strings, plumb all APIs through, etc. There's some current performance issues, namely that small string support introduces branching and our default implementation technique involves spilling to the stack and calling `_UnmanagedString` methods. The first is unavoidable (although we're planning a better overall branching strategy to reduce this impact), the second is TODO.

milseman commented:
@swift-ci please test

(I've been passing clean locally on release+asserts, but not debug+asserts)

milseman commented:
(expecting regressions)

@swift-ci please smoke benchmark

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
Regression from `RemoveWhereQuadraticString` seems to be that small strings added just barely enough code to tilt the inlining heuristic away from one top level of inlining, and thus there's some ARC.

milseman commented:
So yeah, that's not nearly as bad as I feared and I think a lot of it will improve when we switch to a more sane bit-layout and branching scheme.

milseman commented:
Now trying with ASCII literals

@swift-ci please smoke benchmark

milseman commented:
(expected net regression from spill-based approaches, just interested in magnitude)

milseman commented:
Woo, *there's* the regressions I was expecting!

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
(There's some kind of bug that I haven't hunted down yet, so that could also influence benchmark behavior).

@swift-ci please smoke benchmark

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
(added a few sources of dynamic small strings, but haven't done append/join/etc)

@swift-ci please smoke benchmark

milseman commented:
Loads more dynamic small strings. Implementation approach for index-operations is still spill-and-go. Missing some appends still...

@swift-ci please smoke benchmark

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
@swift-ci thank you

milseman commented:
(That last push includes an assert that any ASCII string that can be smol, is)

@swift-ci please smoke benchmark

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
(the only failure I have locally now is StringAPI.swift, which is the hasPrefix on '\0\0' issue).

@swift-ci please test

milseman commented:
(Pushed a big rebase and history rewrite. StringAPI's "\0\0" prefix test is known to be failing until @airspeedswift gets his fix in).

@swift-ci please test

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
@swift-ci please test

milseman commented:
@shahmishal @atrick have you seen this before?

```
https://ci.swift.org/job/swift-PR-Linux/3599/
16:52:09 CMake Error at CMakeLists.txt:90 (message):
16:52:09   LLDB test compilers not specified.  Tests will not run
```

@swift-ci please test linux platform

milseman commented:
@swift-ci please test linux platform

jrose-apple commented:
On another PR Davide suggested doing a clean test to get around the LLDB failure, and it seemed to work.

milseman commented:
Please test with following PR: https://github.com/apple/swift-xcode-playground-support/pull/24

@swift-ci please test

milseman commented:
@swift-ci please smoke benchmark

milseman committed:
[string] Small fast-path for hasPrefix/Suffix

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
So the OS X tests pass locally for me, even after rebasing this morning. Not sure why yesterday had the UXPass.

@swift-ci please test os x platform

milseman commented:
Please test with following PR: https://github.com/apple/swift-corelibs-foundation/pull/1476

@swift-ci please test

milseman commented:
Please test with following PR: apple/swift-corelibs-foundation#1476

@swift-ci please test

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please test os x platform

milseman commented:
Please test with following PR: apple/swift-corelibs-foundation#1484

@swift-ci please test linux platform

milseman commented:
Please test with following PR: apple/swift-corelibs-foundation#1484

@swift-ci please clean test

milseman commented:
(I don't reproduce the simulator crashes locally after having rebased and deleted my build directory, so not sure what's going on here).

milseman commented:
Please test with following PR: apple/swift-corelibs-foundation#1484

@swift-ci please test linux platform

milseman committed:
[string] Initial small string implementation.

This adds a small string representation capable of holding up to 15
ASCII code units directly in registers. This is extendable to UTF-8 in
the future.

It is intended to be the preferred representation whenever possible
for String, and is intended to be a String fast-path. Future small
forms may be added in the future (likely off the fast-path).

Small strings are available on 64-bit, where they are most beneficial
and well accomodated by limited address spaces. They are unavailable
on 32-bit, where they are less of a win and would require much more
hackery due to full address spaces.

milseman committed:
[string] Simplify creation logic, especially for C strings.

Streamline internal String creation. Previously, everything funneled
into a single generic function, however, every single call of the
generic funnel had relevant specific information that could be used
for a more efficient algorithm.

In preparation for efficiently forming small strings, refactor this
logic into a handful of more specialized subroutines to preserve more
specific information from the callers.

milseman committed:
[string] Integrate small strings.

Switch StringObject and StringGuts from opaquely storing tagged cocoa
strings into storing small strings. Plumb small string support
throughout the standard library's routines.

milseman committed:
[string] Aggressively form small strings.

Whenever possible, prefer the small string format. Updates creation
logic. Adds assertion on large strings that they're not intended to be
small.

milseman commented:
Please test with following PR: apple/swift-corelibs-foundation#1484

@swift-ci please test

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
@lorentey please review 😀

milseman commented:
Mac is LLDB failures. @vedantk or @dcci, could you help me investigate tomorrow?

Linux is UXPass. @gottesmm @shajrawi is there a bug tracking this? Can I un-XFAIL it on Darwin with a corresponding bug to investigate?

shajrawi commented:
@milseman the XFAIL / UPASS is tracked by rdar://38152291

gottesmm commented:
@milseman I am working on it this morning.

milseman commented:
Please test with the following PR: https://github.com/apple/swift-lldb/pull/461

@swift-ci please test

milseman commented:
@swift-ci please smoke benchmark

lorentey reviewed (commented):
I can't wait to see this land. 👍

Do you know how you'll deal with big-endian systems? Keeping `_SmallUTF8String._storage` little-endian would work pretty well, I think.

milseman commented on stdlib/public/core/StringObject.swift:
```diff
@@ -387,6 +388,103 @@ extension _StringObject {
+      unsupportedOn32bit()
+#else
+      _sanityCheck(_isSmallUTF8)
+      return rawBits & ~_StringObject._topNibbleMask
```
I may change this, depending on whether there's a tradeoff between masking it off now or on-demand.

milseman commented on stdlib/public/core/StringLegacy.swift:
```diff
@@ -277,3 +294,22 @@ extension String {
+  func _repeated(_ n: Int) -> _StringGuts {
+    _sanityCheck(n > 1)
+    if self._isSmall {
+      // TODO: visitor pattern for something like this...
```
Ideally we will. Otherwise this involves spilling to the stack rather than operating directly in-register.

milseman commented on stdlib/public/core/SmallString.swift:
```diff
@@ -0,0 +1,846 @@
+}
+extension _SmallUTF8String {
+  @_inlineable
+  public // @testable
```
Good idea. It triggers an assert in the sil verifier locally for me. After merging, I'll try on master and file a bug

milseman commented on stdlib/public/core/SmallString.swift:
```diff
@@ -0,0 +1,846 @@
+
+#if arch(i386) || arch(arm)
+#else
+extension _SmallUTF8String {
```
I tried both ways. In general, I found it a little simpler to have the primary interfaces with the `#if` at the top and the implementation at the bottom, where it isn't riddled with `#if`. This may change when we can remove many of these.

milseman commented on stdlib/public/core/StringUnicodeScalarView.swift:
```diff
@@ -181,13 +181,19 @@ extension String {
       internal var _guts: _StringGuts
 
+      // FIXME(TODO: JIRA): the below is absurdly wasteful.
+      // UnicodeScalarView.Iterator should be able to be passed in-registers.
```
I mean, there's no reason `String.UnicodeScalarView.Iterator` should be more than a handful of registers in size. Maybe passing it in registers is not as important if everything calls mutating methods anyways.

milseman commented on stdlib/public/core/String.swift:
```diff
@@ -966,48 +998,26 @@ extension Sequence where Element: StringProtocol {
-    guard let separator = separator else {
+    let capacity = reservation ?? separatorSize
+    var result = ""
+    result.reserveCapacity(capacity)
```
It seems to be so for joined. (I need to pour over and speed up append anyways.)

milseman commented on stdlib/public/core/StringComparison.swift:
```diff
@@ -12,8 +12,27 @@
+}
+
+// HACK: This gets rid of some retains/releases that was slowing down the
+// memcmp fast path for comparing ascii strings. rdar://problem/37473470
```
Good point. I'll try a new PR after this one to remove them

milseman commented:
Please test with the following PR: apple/swift-lldb#461

@swift-ci please test

milseman commented:
Mac failures are new and seem unrelated.

Please test with the following PR: apple/swift-lldb#461

@swift-ci please test OS X Platform

milseman commented:
Mac failures are unrelated, being addressed in https://github.com/apple/swift/pull/15535

@gottesmm, when do you think that you'll be able to look at unused_containers XFAIL? Should I just disable it on Linux for now?

milseman committed:
[test] Temporarily disable UXPASSing Linux test

milseman commented:
Please test with the following PR: apple/swift-lldb#461

@swift-ci please test Linux Platform

milseman commented:
Please test with the following PR: apple/swift-lldb#461

@swift-ci please test OS X Platform

gottesmm commented:
@milseman I would just use a requires line and continue.

milseman commented:
@gottesmm yup https://github.com/apple/swift/pull/14755/commits/bf21a07106ee4f1cdc11014a0e4e2c753fc01022 😉

milseman commented:
@dcci @vedantk so I've verified that my formatting code does handle intermediary nul characters correctly. Do you know why this test might fail? https://ci.swift.org/job/swift-PR-osx/3925/consoleFull#376610391ba62d58e-7248-467b-91e0-c7508d5cf947

When I tried to run LLDB tests locally, I could get all the LIT based ones working but the python ones all kept crashing (likely I had a mismatched version of Python). Are they supposed to work with a CMake based build?

vedantk commented:
@milseman The cmake build of lldb should support running all the python tests. Separately, I can follow up and see why things are failing for you locally. I'll take a look at the test now.

vedantk commented:
@milseman It looks like https://github.com/apple/swift-lldb/pull/461 forwards intermediary nul characters directly to the formatter's output string without escaping them first. I've left a more detailed comment + suggested fix in the PR.

milseman commented:
Please test with the following PR: apple/swift-lldb#461

@swift-ci please test OS X Platform

milseman commented:
Please test with the following PR: apple/swift-lldb#461

@swift-ci please test Linux Platform

milseman commented:
Please test with the following PR: apple/swift-lldb#461
https://github.com/apple/swift-lldb/pull/461
https://github.com/apple/swift-corelibs-foundation/pull/1492

@swift-ci please test Linux Platform

milseman commented:
Please test with the following PR:
apple/swift-lldb#461
apple/swift-corelibs-foundation#1492

@swift-ci please test Linux Platform

milseman commented:
Please test with the following PR:
apple/swift-lldb#461
apple/swift-corelibs-foundation#1492

@swift-ci please test Linux Platform

gottesmm commented:
Congrats Michael!
